### ISSUE TITLE
Select attributes in a self referencing field

### ISSUE THREAD

**Issue by xserrat** [schmittjoh/JMSSerializerBundle#480]
Hi, I tried to serialize only specific fields of an attribute that is a self-referencing relation. I show an example below:

 ```php
use JMS\Serializer\Annotation as JMS;

/**
* @JMS\ExclusionPolicy("all")
*/
class User
{
      /**
      * @JMS\Expose
      *@JMS\Groups({"user-basics", "friends-list"})
      */
      private $id;

      /**
      * @JMS\Expose
      *@JMS\Groups({"user-basics"})
      */
      private $name;
      
      /**
      * @JMS\Expose
      *@JMS\Groups({"user-basics"})
      */
      private $profileImage;
       
      /**
      * @JMS\Expose
      * @JMS\Groups({"user-basics", "friends-list"})
      */
      private $email;

      /**
      * @JMS\Expose
      * @JMS\Groups({"friends-list"})
      */
      private $friends; //self-referencing (User array)
}
```

If I want to obtain about friends, only the id and the email, and other fields of user:

```
{
   "user": {
       "id": 1,
       "name": "Joseph",
       "profileImage": "http://example.com/image.jpg",
       "email": "joseph@example.com",
       "friends": [
              {
                    "id": 1,
                    "email": "friend@example.com"
              }
       ]
   }
}

```
I see that it's not possible because when I try to get data through my controller, I specify the group "user-basics" (to obtain basic info) and the group "friends-list", the controller returns all fields in both (user info and fields).

Some idea?

Thanks.

**Comment by mcornella** [comment]
Maybe you can use the `@MaxDepth` annotation? http://jmsyst.com/libs/serializer/master/cookbook/exclusion_strategies#limiting-serialization-depth-of-some-properties

**Comment by xserrat** [comment]
Hi @mcornella, the ```@MaxDepth``` only works as a delimiter in order to decide what is the depth of a specific field and avoid big depths. Thanks!

**Comment by NicolasCayet** [comment]
+1 for this feature.

Maybe an exclusion with both depth + group strategies can fulfill this ?

**Comment by mcornella** [comment]
Perhaps you could extend the User class into a FriendUser class, annotate that and make the $friends list of class FriendUser instead. 
Haven't tried it, I have no idea if it would work.

**Comment by xserrat** [comment]
I think that it wouldn't work @NicolasCayet  because all strategies will affect to both at the same time... However, as @mcornella says, using another class for user friends could be a good solution. I will try to implement this possibility and I will report results.

**Comment by robmro27** [comment]
Hi I have the same situation do you resolve it somehow ?

**Comment by xserrat** [comment]
Hi @robmro27 ,  

In order to solve the problem I built two entities as @mcornella said like follow:

```php
class User
{
    /**
     * @var integer
     *
     * @ORM\Column(type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     * @JMS\Expose
     * @JMS\Groups({"user-details", "list-users", "user-login", "user-friend-details", "friend-details"})
     */
    private $id;

    /**
     * @var string
     *
     * @ORM\Column(type="string", length=255, nullable=true)
     * @JMS\Expose
     * @JMS\Groups({"user-details", "list-users", "user-login", "user-friend-details","friend-details"})
     */
    private $name;

    /**
     * @var string
     *
     * @ORM\Column(type="string", length=255, nullable=true)
     * @JMS\Expose
     * @JMS\Groups({"user-details", "list-users", "user-login", "user-friend-details","friend-details"})
     */
    private $surname;

    /**
     * @var string
     *
     * @ORM\Column(type="string", length=255, unique=true)
     * @JMS\Expose
     * @JMS\Groups({"user-details", "list-users", "user-login", "user-friend-details","friend-details"})
     */
    private $email;

    /**
     * @ORM\OneToMany(targetEntity="UserLink", mappedBy="user")
     * @JMS\Expose
     * @JMS\Groups({"user-login", "user_link-friend-status", "user-details", "list-friends"})
     * @JMS\MaxDepth(3)
     */
    private $friends;

    // ...
}
```

```php
class UserLink
{
    const STATUS_PENDING_ACCEPT = 1;
    const STATUS_PENDING_CONFIRM = 2;
    const STATUS_LINKED = 3;

    /**
     * @ORM\Column(type="integer")
     * @ORM\Id
     * @ORM\GeneratedValue(strategy="AUTO")
     */
    private $id;

    /**
     * @ORM\ManyToOne(targetEntity="User", inversedBy="friends")
     * @ORM\JoinColumn(name="user_id", referencedColumnName="id")
     */
    private $user;

    /**
     * @ORM\ManyToOne(targetEntity="User")
     * @ORM\JoinColumn(name="partner_id", referencedColumnName="id")
     * @JMS\Expose
     * @JMS\Groups({"user_link-friend-status", "user-friend-details"})
     */
    private $friend;

    /**
     * @ORM\Column(type="integer")
     * @JMS\Expose
     * @JMS\Groups({"user_link-friend-status"})
     */
    private $status;

    public function __construct(User $user, User $friend, $status)
    {
        // ...
    }
}
```

So, using this strategy, we have a User with all properties you need like `name`, `surname`, `email` and a UserLink collection called `$friends`. Having a UserLink, we can link each User with his partner and the relation with both (if you want to save some status when the user accepts another or something like this).

To sum up, if you have another entity in order to link a user with another one, you'll be able to define different groups to obtain the properties you want.

I expect that it solves your doubts!

**Comment by bvisonl** [comment]
In case anyone is still looking into this... for my specific case when serializing the parent property I am only interested in knowing 2 properties of the child self reference... so what I did was that I created a virtual property that returned an array containing those 2 properties and I added the serialization group to that virtual property.. so the self reference won't get called again.
